### DISCUSSION TITLE
Laravel Queue Slow Query

### DISCUSSION THREAD

**Discussion by ekaterina4952** [laravel/framework#50144] [Q&A] [answered]
Hi, I am using laravel queue. but I am struggling with so slowing on laravel application. (CPU 8, RAM 32GB, IOPS 100MB/s)
Most slow query log are related with queue. over 2k slow query log per day. Please advise. Thanks advance.

select * from jobs where queue = 'default' and ((reserved_at is null and available_at <= 1707701703) or (reserved_at <= 1707701613)) order by id asc limit 1 FOR UPDATE SKIP LOCKED;
Time: 2024-02-12T01:35:42.568834Z
User@Host: api[api] @ localhost [127.0.0.1]  Id: 71261
Query_time: 38.893383  Lock_time: 0.000002 Rows_sent: 0  Rows_examined: 11

**Comment by ekaterina4952** [comment]
Laravel Pulse

![image](https://github.com/laravel/framework/assets/87656984/44d192f4-2b05-467b-a838-aa3ec52c4e5d)

**Reply by ekaterina4952** [reply]
![image](https://github.com/laravel/framework/assets/87656984/b566a6c2-7199-4915-adce-06fe146da3a3)

**Comment by ghost** [comment, marked as answer]
DB queue is not recommended in prod environment because its inability to offer ATOMIC locks. SQS or other messaging system is recomended.

**Comment by ekaterina4952** [comment]
Hi @marius-mcp, this query slowing happens too many per day. any thoughts? Thanks in advance.

![image](https://github.com/user-attachments/assets/97212131-bf8b-4e26-865c-f67b5e6f1957)

**Reply by ghost** [reply]
Put a composed index on deleted_at and id if you don't have yet one.

**Reply by ekaterina4952** [reply]
Thanks for your update, however, id is primary key and laravel says no need index regarding deleted_at. https://stackoverflow.com/questions/26525068/does-laravels-soft-delete-need-index-on-mysql/26530488#26530488

**Reply by ghost** [reply]
Laravel says that lumen is not to be used, also that octane could be used. Don't trust all laravel says :)

**Reply by ghost** [reply]
Put the index, test and then if you don't like it remove the index.

**Reply by ekaterina4952** [reply]
delete_at index isnt helpful for me

**Reply by ghost** [reply]
Update
@ekaterina4952  Please run explain and then your query and show the results (print screen).

**Reply by ghost** [reply]
@ekaterina4952 see this issue https://github.com/laravel/framework/issues/54068

**Reply by ekaterina4952** [reply]
`explain select * from `table` where `id` = 123445678 and `table`.`deleted_at` is null limit 1`

![image](https://github.com/user-attachments/assets/c289d796-595c-4f8e-aa35-9d05e3838f41)

**Reply by ghost** [reply]
@ekaterina4952 123 mil rows is something I don't have experienced (my exp is limited to few millions only) so I can only guess that 5 seconds to fetch a row by its primary key must be because of server business or row/table lock. Anyway thanks for this as I saw that a composed index on deleted_at and another column is not used when deleted_at is nullable.
